tests/models/decoder_only/vision_language/utils.py

```diff
@@ -0,0 +1,999 @@
+    if size_wrapper.type != SizeType.SIZE_FACTOR or not \
+            all(factor == 1.0 for factor in size_wrapper.data):
+        raise ValueError("Embedding tests require constant (1.0) size factors")
+    if test_info.convert_assets_to_embeddings is None:
+        raise ValueError("No conversion func for getting embeddings found")
+
+    model_prompts = get_model_prompts(
+        SINGLE_IMAGE_BASE_PROMPTS,
+        test_info.img_idx_to_prompt,
+        test_info.video_idx_to_prompt,
+        test_info.prompt_formatter,
+    )
+
+    images = [asset.pil_image for asset in image_assets]
+    embeds = test_info.convert_assets_to_embeddings(image_assets)
+    assert len(images) == len(model_prompts)
+
+    inputs = build_single_image_inputs(images, model_prompts, size_wrapper)
+    vllm_embeddings = build_single_image_inputs(embeds, model_prompts,
+                                                size_wrapper)
+    return inputs, vllm_embeddings
+
+
+def build_video_inputs_from_test_info(
+    test_info: VLMTestInfo,
+    video_assets: _VideoAssets,
+    size_wrapper: ImageSizeWrapper,
+    num_frames: int,
+):
+    if test_info.prompt_formatter is None:
+        raise ValueError("Prompt formatter must be set to build video inputs")
+    model_prompts = get_model_prompts(
+        [VIDEO_BASE_PROMPT],
+        test_info.img_idx_to_prompt,
+        test_info.video_idx_to_prompt,
+        test_info.prompt_formatter,
+    )
+
+    sampled_vids = [
+        sample_frames_from_video(asset.np_ndarrays, num_frames)
+        for asset in video_assets
+    ]
+
+    video_scaler = (resize_video if size_wrapper.type == SizeType.FIXED_SIZE
+                    else rescale_video_size)
+
+    return [(
+        [prompt for _ in size_wrapper.data],
+        [video_scaler(video, size) for size in size_wrapper.data],
+    ) for video, prompt in zip(sampled_vids, model_prompts)]
+
+
+def apply_image_size_scaling(image, size: Union[float, Tuple[int, int]],
+                             size_type: SizeType):
+    """Applies a size scaler to one image; this can be a an image size factor,
+    which scales the image while maintaining the aspect ratio"""
+    # Special case for embeddings; if it's a tensor, it's only valid if we
+    # are considering size factors at constant scale, i.e., we just clone
+    # the tensor
+    if isinstance(image, torch.Tensor):
+        assert size_type == SizeType.SIZE_FACTOR and size == 1
+        return image
+    if size_type == SizeType.SIZE_FACTOR:
+        # We have a list of image size factors
+        return rescale_image_size(image, size)
+    elif size_type == SizeType.FIXED_SIZE:
+        # We have a list of fixed sizes
+        return image.resize(size)
+    raise ValueError("ImageSizeWrapper type must be FIXED_SIZE or SIZE_FACTOR")
+
+
+####### Entrypoints for running different test types
+# Wrappers for all the above, where the only difference
+# is that each test runs in a separate process
+def run_single_image_test(
```

Comment:
Is there an easy way to just run the tests for a single model, like the case for PP?

Reply:
Good point - I don't think there is at the moment, but I think it can probably be done by setting IDs in the parametrize call using the model key as a prefix. I'll give try it this evening!

Reply:
I looked into a bit - I think the best way to do this is through substring matching.

The reason for this is that this is how an expanded test case looks after parametrizing:
```
test_single_image_models[intern_vl-OpenGVLab/InternVL2-1B-128-5-half-None-size_wrapper19]
```

So we can filter that with substrings (`-k`). The key in the settings is always the first parameter in every test, so we can leverage `[<key_name>` as a matchable substring for running all of the tests for a specific architecture. Some examples:

- Runs all of the types of tests for only llava - since model architectures use `_`, the hyphen after `llava` prevents it from matching against the `llava_next` tests
```bash
pytest test_models.py -k  "[llava-"
```

- Runs only the tests for one of the models for InternVL (which runs its tests against two models)
```bash
pytest test_models.py -k OpenGVLab/InternVL2-1B
```

You can also use `and/or` inside of `-k` filters, so you can also do stuff like this to run just the single image llava tests, to run tests for multiple model types, etc. Note that this also matches `test_single_image_models_heavy`, so we don't need to worry about which tests are distributed when testing like this
```
pytest test_models.py -k "test_single_image and [llava-"
```


I added a long comment with similar examples, and recommended  using `--collect-only` to quickly check which tests pytests will run with a command before actually executing it here: https://github.com/vllm-project/vllm/blob/f9f9fc927a1b7bc9f3f9d878807a7fa7fc78581c/tests/models/decoder_only/vision_language/test_models.py#L46